File: epi_judge_python_solutions/longest_subarray_with_distinct_values.py
from test_framework import generic_test


def longest_subarray_with_distinct_entries(A):

    # Records the most recent occurrences of each entry.
    most_recent_occurrence = {}
    longest_dup_free_subarray_start_idx = result = 0
    for i, a in enumerate(A):
        # Defer updating dup_idx until we see a duplicate.
        if a in most_recent_occurrence:
            dup_idx = most_recent_occurrence[a]
            # A[i] appeared before. Did it appear in the longest current
            # subarray?
            if dup_idx >= longest_dup_free_subarray_start_idx:
                result = max(result, i - longest_dup_free_subarray_start_idx)
                longest_dup_free_subarray_start_idx = dup_idx + 1
        most_recent_occurrence[a] = i
    return max(result, len(A) - longest_dup_free_subarray_start_idx)


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main(
            "longest_subarray_with_distinct_values.py",
            'longest_subarray_with_distinct_values.tsv',
            longest_subarray_with_distinct_entries))

File: epi_judge_cpp_solutions/longest_subarray_with_distinct_values.cc
#include <algorithm>
#include <unordered_map>
#include <vector>

#include "test_framework/generic_test.h"

using std::max;
using std::unordered_map;
using std::vector;

int LongestSubarrayWithDistinctEntries(const vector<int>& A) {
  // Records the most recent occurrences of each entry.
  unordered_map<int, size_t> most_recent_occurrence;
  size_t longest_dup_free_subarray_start_idx = 0, result = 0;
  for (size_t i = 0; i < size(A); ++i) {
    const auto& [inserted_entry, inserted_happen] =
        most_recent_occurrence.emplace(A[i], i);
    // Defer updating dup_idx until we see a duplicate.
    if (!inserted_happen) {
      // A[i] appeared before. Did it appear in the longest current subarray?
      if (inserted_entry->second >= longest_dup_free_subarray_start_idx) {
        result = max(result, i - longest_dup_free_subarray_start_idx);
        longest_dup_free_subarray_start_idx = inserted_entry->second + 1;
      }
      inserted_entry->second = i;
    }
  }
  return max(result, size(A) - longest_dup_free_subarray_start_idx);
}

// clang-format off


int main(int argc, char* argv[]) {
  std::vector<std::string> args {argv + 1, argv + argc};
  std::vector<std::string> param_names {"A"};
  return GenericTestMain(args, "longest_subarray_with_distinct_values.cc", "longest_subarray_with_distinct_values.tsv", &LongestSubarrayWithDistinctEntries,
                         DefaultComparator{}, param_names);
}
// clang-format on
